docs/src/cookbooks/hierarchy.md

```diff
@@ -167,7 +170,92 @@ val chiselCircuit = (new chisel3.stage.phases.Elaborate)
 println("```")
 ```
 
-## How do I look up parameters from a Definition, if I don't want to instantiate it?
+## How do I make case classes containing Data or Modules accessible from an instance?
+
+For case classes containing `Data`, `BaseModule`, `MemBase` or `Instance` types, you can provide an implementation of the `Lookupable` typeclass.
+
+**Note that Lookupable for Modules is deprecated, please cast to Instance instead (with `.toInstance`).**
+
+Consider the following case class:
+
+```scala mdoc:reset
+import chisel3._
+import chisel3.experimental.hierarchy.{Definition, Instance, instantiable, public}
+
+case class UserDefinedType[T <: Data, M <: RawModule](name: String, data: T, inst: Instance[M])
+```
+
+By default, instances of `UserDefinedType` will not be accessible from instances:
+
+```scala mdoc:fail
+class ChildModule extends Module {
+  @public val wire = Wire(UInt(8.W))
+}
+@instantiable
+class HasUserDefinedType extends Module {
+  val inst = Module(new ChildModule)
+  val wire = Wire(UInt(8.W))
+  @public val x = UserDefinedType("foo", wire, inst.toInstance)
+}
+```
+
+We can implement the `Lookupable` type class for `UserDefinedType` in order to make it accessible.
+This involves defining an implicit def in the companion object for `UserDefinedType`.
+Because `UserDefinedType` has three fields, we use the `Lookupable.product3` factory.
+It takes 4 type parameters: the type of the case class, and the types of each of its fields.
+It also takes two functions: one to convert the case class to a tuple of its fields, and one to convert the tuple back to the case class.
+
+**If any fields are `BaseModules`, you must change them to be `Instance[_]` in order to define the `Lookupable` typeclass.**
+
+For more information about typeclasses, see the [DataView section on Type Classes](https://www.chisel-lang.org/chisel3/docs/explanations/dataview#type-classes).
+
+```scala mdoc
+import chisel3.experimental.hierarchy.Lookupable
+object UserDefinedType {
+  // Use Lookupable.Simple type alias as return type.
+  implicit def lookupable[T <: Data, M <: RawModule]: Lookupable.Simple[UserDefinedType[T, M]] =
```

Comment:
it feels like this example is more confusing than necessary with the `T` and `M` type parameters. Could you show a simpler example first with concrete types like `MyBundle` and `MyModule`? It's hard to separate in my mind which part of this is type parameters for the `Lookupable.product3` and which are for the fact that our UserDefinedType is type parameterized itself.

Reply:
Yeah I agree we should have a dead simple example, and then a second one with the type parameters.

Reply:
Done, I split it into 3 sections, a simpler example, the type parameterized one, and then the "lots of fields" is its own.